Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -4,34 +4,230 @@
     int value;
-} Hash;
+} Item;
 
-void print() {
+//create global testHash and expectedHash for testFunctions to use
+Item* testHash[SIZE];
+Item* expectedHash[SIZE];
 
+// traverse function is for visual confirmation of differences between hashTables when building functionality. 
+// void traverse() {
+//     for (int i = 0; i < SIZE; i++) {
+//         int value = testHash[i]->value;
+//         int key = testHash[i]->key;        
+//         printf("Test Key: %d Value: %d \n", key, value);
+//     }
+//     for (int i = 0; i < SIZE; i++) {
+//         int value = expectedHash[i]->value;
+//         int key = expectedHash[i]->key;
+//         printf("Expected Key: %d Value: %d \n", key, value);
+//     }
+// }
+
+//compare checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool compare() {
+    for (int i = 0; i < SIZE; i++) {
+        int testKey = testHash[i]->key;
+        int testValue = testHash[i]->value;
+        int expectedKey = expectedHash[i]->key;
+        int expectedValue = expectedHash[i]->value;
+        if (testKey != expectedKey || testValue != expectedValue) {
+            return false;
+        }
+    }
+    return true;
 }
 
-void traverse() {
+// hashing takes in a key
+// returns a "hashed" index based on the SIZE of the hashTable
+int hashing(int key) {
+    return key % SIZE;
+}
 
+// creates a new Item struct
+// returns new struct
+Item * createItem(int key, int value, int * success) {
+    Item * item = malloc(sizeof(Item));
+    if (item != NULL) {        
+        item->key = key;
+        item->value = value;
+        *success = SUCCESS;
+        return item;
+    }
 }
 
-bool compare() {
-    return true;
-    return false;
+// insert takes in a key and value
+// calls a helper function to create an Item struct with the provided key and value
+// adds the new Item to the hashTable
+// if the expected value is in the hashTable at the expected key then returns a success code
+int insert(int type, int key, int value) {
+    int success = ERROR;
+    if (type == TEST) {
+        testHash[key] = createItem(key, value, &success);
+        if (testHash[key]->value != value) {
+            return ERROR;
+        }
+        
+    } else if (type == EXPECTED) {
+        expectedHash[key] = createItem(key, value, &success);
+        if (expectedHash[key]->value != value) {
+            return ERROR;
+        }
+    }
+    return SUCCESS;
+}
+
+// bulkInsert receives an array of keys and an array of values
+// key[i] will be mapped to value[i] when building the hashTable
+// loops through the values to create a key for the current value then inserts it into the existing hashTable.
+// returns a SUCCESS code if all values are added into the hashTable successfully 
+int bulkInsert(int type, int *keys, int *values) {
+    for (int i = 0; i < SIZE; i++) {
+        int key = hashing(keys[i]);
+        return insert(type, key, values[i]);
```

Comment:
Returning here will result in returning after the first call to `insert`.

Reply:
Thanks for catching! Updated in recent [commit 7b759b3c93b27fb6aebfae99f124f8559fcf28af](https://github.com/kcarrel/Training-Notes/pull/12/commits/7b759b3c93b27fb6aebfae99f124f8559fcf28af)